src/garage/envs/environment.py

```diff
@@ -0,0 +1,188 @@
+"""Base Garage Environment API."""
+
+import abc
+
+
+class Environment(abc.ABC):
+    """The main API for garage environments.
+
+    The main public API methods are:
+        * step()
+        * reset()
+        * render()
+        * visualize()
+        * close()
+
+    Set the following attributes:
+        * action_space: The action space specification
+        * observation_space: The observation space specification
+        * spec: The environment specifications
+        * render_modes: The list of supported render modes
+
+    Make sure your environment is pickle-able:
+        Garage pickles the environment via the `cloudpickle` module
+        to save snapshots of the experiment. However, some environments may
+        contain attributes that are not pickle-able (e.g. a client-server
+        connection). In such cases, override `__getstate__()` and
+        `__setstate()__` to add your custom pickle logic.
+
+    """
+
+    @property
+    @abc.abstractmethod
+    def action_space(self):
+        """Return the action space specification.
+
+        Some environments do not provide their action space specifications as
+        type akro.Space. In such cases, subclasses should silently convert
+        the environment's action_space to akro.Space.
+
+        Returns:
+            akro.Space: The envionrment specification.
+
+        """
+
+    @property
+    @abc.abstractmethod
+    def observation_space(self):
+        """Return the observation space specification.
+
+        Some environments do not provide their observation space
+        specifications as type akro.Space. In such cases, subclasses should
+        silently convert the environment's observation_space to akro.Space.
```

Comment:
"silently" converts no longer appies -- the types here are akro.Space, full stop.